File: leetcode/Python/next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        sol, st, um = list(), list(), dict()
        for n in nums2:
            while st and st[-1]<n:
                um[st.pop()]=n
            st.append(n)
        while st:
            um[st.pop()]=-1
        for n in nums1:
            sol.append(um[n])
        return sol

File: leetcode/C++/next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> um;
        stack<int> st;
        vector<int> sol;
        for(int &n : nums2){
            while(!st.empty() && st.top()<n){ um[st.top()]=n; st.pop();}
            st.push(n);
        }
        while(!st.empty()){ um[st.top()]=-1; st.pop();}
        for(int &n : nums1) sol.push_back(um[n]);
        return sol;
    }
};